### PR TITLE
Add array_average Presto function

### PR DESCRIPTION
**Pull request by jwyles-ahana**
Implementing array_average as a simple function.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@jwyles-ahana Thank you for adding one more Presto function. Please, add documentation and look into CI failures.

File: velox/functions/prestosql/ArrayFunctions.h [hunk 1]
```diff
@@ -316,4 +316,41 @@ struct ArraySumFunction {
...
+      }
+    }
+    if (count != 0) {
+      out = sum / count;
+    }
+    return count != 0;
+  }
+
+  template <typename TOutput, typename TInput>
+  FOLLY_ALWAYS_INLINE void callNullFree(TOutput& out, const TInput& array) {
```
Comment (kevinwilfong) [inline velox/functions/prestosql/ArrayFunctions.h:339]: To be consistent with the implementation of "call" and the Presto documentation this should return null if the input array is empty, so the return type should be bool.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      const std::vector<VectorPtr>& input) {
+    auto result = evaluate<FlatVector<T>>(expression, makeRowVector(input));
+    assertEqualVectors(expected, result);
+  }
+};
+
+} // namespace
+
+// Test floating point arrays
+TEST_F(ArrayAverageTest, doubleInput) {
```
Comment (kevinwilfong) [inline velox/functions/prestosql/tests/ArrayAverageTest.cpp:42]: Could you add a case for empty array as well.

And also add a case where all of the arrays have non-null elements (including an empty array again), to exercise the callNullFree code path.

**Review by kevinwilfong — COMMENTED** [review]
Thanks for contributing this!

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp [hunk 3]
```diff
@@ -30,6 +30,11 @@ inline void registerArraySumFunction() {
   registerFunction<ArraySumFunction, OT, IT>({"array_sum_alt"});
 }
 
+template <typename OT, typename IT>
```
Comment (aditi-pandit) [inline velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:33]: Nit: Since we don't bind this function for multiple types, the usage of template variables doesn't seem worthwhile to me.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  auto input = makeNullableArrayVector<double>(
+      {{0.0, 1.0, 2.0}, {std::nullopt, 1.0, 2.0}, {std::nullopt}});
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArrayAverageTest.cpp:44]: Please can you add test cases for corner cases like std::numeric_limits<T>::infinity(), min(), max(), quiet_Nan...etc.

**Review by aditi-pandit — COMMENTED** [review]
Thanks for adding this function. Please add documentation in array.rst.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/functions/prestosql/tests/ArrayAverageTest.cpp:44]: +1

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp (hunk 3 above)
Comment (mbasmanova) [reply velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:33]: +1

File: velox/functions/prestosql/ArrayFunctions.h [hunk 5]
```diff
@@ -316,4 +316,40 @@ struct ArraySumFunction {
   }
 };
 
+template <typename T>
+struct ArrayAverageFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+  template <typename TOutput, typename TInput>
+  FOLLY_ALWAYS_INLINE bool call(TOutput& out, const TInput& array) {
+    TOutput sum = 0;
+    int count = 0;
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:325]: nit: int64_t or size_t?

File: velox/functions/prestosql/ArrayFunctions.h [hunk 6]
```diff
@@ -316,4 +316,40 @@ struct ArraySumFunction {
   }
 };
 
+template <typename T>
+struct ArrayAverageFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+  template <typename TOutput, typename TInput>
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:322]: TOutput is always double.

File: velox/functions/prestosql/ArrayFunctions.h (hunk 1 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/ArrayFunctions.h:339]: Change made.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp (hunk 2 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/tests/ArrayAverageTest.cpp:42]: Tests added.

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp (hunk 3 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:33]: I have removed the registerArrayAverageFunction() method.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp (hunk 4 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/tests/ArrayAverageTest.cpp:44]: Test added.

File: velox/functions/prestosql/ArrayFunctions.h (hunk 5 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/ArrayFunctions.h:325]: Changed to size_t

File: velox/functions/prestosql/ArrayFunctions.h (hunk 6 above)
Comment (jwyles-ahana) [reply velox/functions/prestosql/ArrayFunctions.h:322]: Rewrote call and callNullFree methods as non-templated.

**Comment by jwyles-ahana** [comment]
@mbasmanova @laithsakka What else needs to be done for this PR?

File: velox/functions/prestosql/ArrayFunctions.h [hunk 7]
```diff
@@ -316,4 +316,40 @@ struct ArraySumFunction {
...
+      out = sum / count;
+    }
+    return count != 0;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullFree(
+      double& out,
+      const arg_type<Array<double>>& array) {
+    // If the array is empty then set result to null
+    if (array.size() == 0)
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:343]: add {
}

**Review by laithsakka — APPROVED** [review]
looks good just one minor missing braces before i import it.

File: velox/functions/prestosql/tests/ArrayAverageTest.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  auto expected = makeNullableFlatVector<double>({1.0, 1.5, std::nullopt});
+  testExpr<double>(expected, "array_average(C0)", {input});
+}
+
+// Test floating point arrays: limits cases
+TEST_F(ArrayAverageTest, doubleLimitsInput) {
+  using limits = std::numeric_limits<double>;
+  auto input = makeNullableArrayVector<double>(
+      {{limits::infinity(), 1.0, 2.0},
+       {limits::min(), 1.0, 2.0},
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayAverageTest.cpp:54]: @jwyles-ahana There is a linter warning. Would you double check?

```
Use of std::numeric_limits<T>::min() on floating-point type T.
Use std::numeric_limits<T>::lowest() instead ofstd::numeric_limits<T>::min() where T is a floating-point type.
std::numeric_limits<T>min() behaves counter-intuitively when T isa floating-point type (e.g. float, double, long double, etc.). Thismost likely does not do what the author intends.
Lint code: CLANGTIDY
Lint name: facebook-hte-FloatingPointMin
-       {limits::min(), 1.0, 2.0},
+       {limits::lowest(), 1.0, 2.0},
```

**Comment by mbasmanova** [comment]
@jwyles-ahana This PR is almost complete, but there is a linter warning. Any chance you could fix it so we could merge this PR?

**Comment by jwyles-ahana** [comment]
@mbasmanova @laithsakka I have made the requested changes and now all checks are passing except the conbench one. I am not sure how anything in this commit could affect any of the existing benchmarks... Any ideas on what to look for to fix this?

**Comment by mbasmanova** [comment]
@jwyles-ahana Thanks. conbench is known to be flaky. It is fine to ignore the failures. CC: @pedroerp @raulcd @kgpai 

@laithsakka Laith, would you re-import and land this PR?

**Comment by mbasmanova** [comment]
@jwyles-ahana @laithsakka Folks, anything blocking this PR?

**Comment by kgpai** [comment]
@jwyles-ahana Can you update with latest ? Should be good to merge/land after that .

**Comment by isadikov** [comment]
@aditi-pandit I can take over the function. Maybe we can close this PR and I will open a new one with the same changes (I cannot push directly to this fork).

**Comment by aditi-pandit** [comment]
> @aditi-pandit I can take over the function. Maybe we can close this PR and I will open a new one with the same changes (I cannot push directly to this fork).

@isadikov : Yes, that is fine. We had contacted James, but didn't hear back. Please refernce this original PR in your new one.

**Comment by isadikov** [comment]
I am still working on this.
